File: src/omnia/data/structure/observable/writable/GenericMutationEvent.java
package omnia.data.structure.observable.writable;

import omnia.data.structure.Collection;
import omnia.data.structure.observable.ObservableDataStructure;

class GenericMutationEvent<StateType, OperationsType extends Collection<?>>
    implements ObservableDataStructure.MutationEvent {
  private final StateType state;
  private final OperationsType operations;

  protected GenericMutationEvent(StateType state, OperationsType operations) {
    this.state = state;
    this.operations = operations;
  }

  @Override
  public StateType state() {
    return state;
  }

  @Override
  public OperationsType operations() {
    return operations;
  }
}

File: src/omnia/data/structure/observable/writable/GenericMutationEvent.kt
package omnia.data.structure.observable.writable

import omnia.data.structure.Collection
import omnia.data.structure.observable.ObservableDataStructure

internal open class GenericMutationEvent<StateType : Any, OperationsType : Collection<*>> protected constructor(private val state: StateType, private val operations: OperationsType) : ObservableDataStructure.MutationEvent {
    override fun state(): StateType {
        return state
    }

    override fun operations(): OperationsType {
        return operations
    }
}